### PR TITLE
Fix #1559: Improved Java 8 compatibility when parsing IEEE754 strings.

### LINKED ISSUE #1559: toFloat/toDouble throw `NumberFormatException` when given trailing designation.
Implicit conversion  toFloat/toDouble from String crashes with java.lang.NumberFormatException if you append float or double 'd' , 'D' , 'f' , 'F' eg. "23423.2323d".toDouble . "23423.2323D".toDouble, "23423.2323f".toDouble
i have developed a precheck to test if a String loosely conforms to decimal number:
Decimal checking is 
```scala
object isDecimal {

  def apply(in: String): Boolean = {
    val len = in.length
    def decimal_(in: String, index: Int,prevWasDot: Boolean,seenE: Boolean): Boolean =
      if(index == len) true //all is good

      else in.charAt(index) match {

	case digit if digit.isDigit =>
	  decimal_(in, index + 1,prevWasDot,seenE)

	case '.' if !prevWasDot =>
	  decimal_(in, index + 1,prevWasDot = true,seenE)

	//this is the case when double is "2342.e2" which is fine
	case e if (e == 'e' | e == 'E') && !seenE && prevWasDot && index + 1 < len && in.charAt(index + 1).isDigit  =>
	  decimal_(in, index + 1,prevWasDot,seenE = true)

	case e if (e == 'e' | e == 'E') && !prevWasDot && !seenE =>
	  decimal_(in, index + 1,prevWasDot,seenE = true)

	// "23.f" "2324.d" are valid decimals but "2323.ed" "2323.ef" are not.
	case last if index == (len - 1)
	  && (last == 'd' | last == 'D' | last == 'f' | last == 'F')
	  && in.charAt(index - 1) != 'e'
	  && in.charAt(index - 1) != 'E' => true

	case sign if index == 0 && (sign == '+' | sign == '-' ) =>
	  decimal_(in, index + 1,prevWasDot,seenE)

	case _ => false
      }
    
      if(in == "Infinity" | in == "-Infinity" | in == "NaN" | in == "-NaN") true else
      
      len match {
	case 0 => false
	case 1 if !in.charAt(0).isDigit => false
	case 2 if in.charAt(0) == '.' && !in.charAt(1).isDigit => false
        case e if e > 0 && in.charAt(0) == 'e' => false
	case _ => decimal_(in,0,prevWasDot = false,seenE = false)
      }
  }
}
```

### PR DESCRIPTION
**Pull request by LeeTibbert**
* This PR fixes Issue #1559 "toFloat/toDouble throw `NumberFormatException`
    when given trailing designation".

    My thanks to @teodimoff for detecting & reporting this issue.

    Strings such as "3.14f".toDouble and "3.14D".toFloat are now accepted in
    the same way they are in Java 8.

  * Also fixes, in passing, Scala Native PR #1836. Added test case from that Issue.

  * This PR supersedes PR #1638, which I will now close.  I had forgotten
    that I had already fixed the issue two months ago, duh!

    Upon comparing the code, I like the unified parseIEEE754 approach better
    because it is DRY.

    I rolled the more Java 8 compatible exception message from PR #1638
    forward into this one.

    The parseIEEE754 code handles close-to-zero and +/- Infinity
    better.

  * Java 8 allows contiguous whitespace in the after/to_the_left_of
    the IEEE754 value. This whitespace can be blanks, tabs, newlines
    or any character between 0 (NUL) and 0x20 (space).

  * I also fixed the way the parse handled values too close to zero
    or too far from zero in either positive or negative directions.
    These cases now return the expected values of 0.0, +Infinity, and
    -Infinity, respectively.

  * Any NumberFormatException thrown now uses the Java 8 idiom of
    echoing the input string within quotes: 'For input string: "+9z.93"'.
    This allows one to detect trailing whitespace.

 * Test cases were added to DoubleSuite.scala & FloatSuite.scala for
   the conditions covered by this PR. Test cases for parsing
   hexadecimal strings were also added to improve coverage.

Documentation:

  * The standard changelog entry is requested.

Testing:

  * Built and tested ("test-all") in Debug mode using sbt 1.3.12 on
    X86_64 only . All tests pass.

### CODE REVIEW

**Comment by LeeTibbert** [comment]
Travis CI appears to be having a bad day, so I am having a bad day.
I have restarted my job twice, so am giving up for awhile (days).

To the best that I can tell the failures are internal to Travis and have
nothing to do with this PR beyond the fact that this is what was executing.

**Comment by LeeTibbert** [comment]
rebased & Travis CI is all green.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  def parseIEEE754[T](s: String, f: (CString, Ptr[CString]) => T): T = {
+    Zone { implicit z =>
+      val cstr = toCString(s)
+      val end  = stackalloc[CString]
+
+      errno.errno = 0
+      var res = f(cstr, end)
+
+      def exceptionMsg = s"For input string \042${s}\042"
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:45]:
```suggestion
      def exceptionMsg = s"For input string \"$s\""
```

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+      if (errno.errno != 0) {
+        if (errno.errno != ERANGE) {
+          // The need to use \042 for double quote seems to be a Scala 2.11 bug.
+          // Uglier workarounds exist.
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:50]: This comment here does not make sense. Should it apply to `def exceptionMsg` instead? If yes, is it still relevant?

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          throw new NumberFormatException(exceptionMsg)
+        }
+        // Else strtod() or strtof() will have returned the proper type for
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:53]: "returned the proper type"? I don't understand.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import scalanative.posix.errno.ERANGE
+
+private[java] object IEEE754Helpers {
+
+  // Java parseDouble() and parseFloat() allow characters at & after
+  // after where C strtod() or strtof() stops. Continuous trailing whitespace
+  // with or without an initial 'D', 'd', 'F', 'f' size indicator
+  // is allowed.
+  // The whitespace can include tabs ('\t') and even newlines! ('\n').
+  // Perhaps even Unicode....
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:16]: "Perhaps" is not very precise.

The JavaDoc has a very clear definition of what's allowed at https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html#valueOf-java.lang.String-
There is even a regex for valid inputs, which we also use in Scala.js:
https://github.com/scala-js/scala-js/blob/0e8fe18808a10887db073cf33c6211721a064e5b/javalanglib/src/main/scala/java/lang/Double.scala#L75-L85
so whitespace chars are the ones in the range `\u0000` to `\u0020` included, and only those.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      true
+    } else {
+      val rest = tail.tail.trim
+      (tail(0): @scala.annotation.switch) match {
+        case 'D' | 'd' => vetIEEE754Tail(rest)
+        case 'F' | 'f' => vetIEEE754Tail(rest)
+        case _         => (tail.trim.length == 0)
+      }
+    }
+  }
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:35]: Since we know that chars in the range [0-32] are admissible, we can easily write this without `trim`, not causing any more allocation than necessary:
```scala
private def vetIEEE754Tail(tail: String): Boolean = {
  val len = tail.length
  if (len == 0) true
  else {
    var i = 0
    if ((tail.charAt(0) & 0xdd) == 0x44) // magic: is first char one of D d F f
      i = 1
    while (i != len) {
      if ((tail.charAt(i) & 0xff) > ' ')
        return false
      i += 1
    }
    true
  }
}
```
And to avoid even allocating the substring at call site, we can take the start index instead of assuming that we have to start at 0:
```scala
private def vetIEEE754Tail(tail: String, start: Int): Boolean = {
  val len = tail.length
  if (len == start) true
  else {
    var i = start
    if ((tail.charAt(i) & 0xdd) == 0x44) // magic: is first char one of D d F f
      i += 1
    while (i != len) {
      if ((tail.charAt(i) & 0xff) > ' ')
        return false
      i += 1
    }
    true
  }
}
```

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+          // The need to use \042 for double quote seems to be a Scala 2.11 bug.
+          // Uglier workarounds exist.
+          throw new NumberFormatException(exceptionMsg)
+        }
+        // Else strtod() or strtof() will have returned the proper type for
+        // 0.0 (too close to zero) or +/- infinity. Slick C lib design!
+      } else if (!end == cstr) {
+        throw new NumberFormatException(exceptionMsg)
+      } else if ((!end(0) == 0) ||
+                 vetIEEE754Tail(s.slice((!end - cstr).toInt, s.length))) {
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:58]: Testing that `!end(0) == 0` is not correct, because there might be a `\u0000` followed by non-whitespace characters in the original string. With the suggested definition of `vetIEEE754Tail`, we can simply do:
```suggestion
      } else if (vetIEEE754Tail(s, (!end - cstr).toInt)) {
```

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        res
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/IEEE754Helpers.scala:59]: This is dead code. We return `res` below. You can leave an empty block, or, better, invert the condition of the `else if` just above.

File: unit-tests/src/test/scala/java/lang/DoubleSuite.scala [hunk 8]
```diff
@@ -195,10 +195,58 @@ object DoubleSuite extends tests.Suite {
...
+    val epsilon = 0.0001
+    assert(Math.abs(Double.parseDouble("6.66D") - 6.66) < epsilon, "a8")
+
+    // Java allows trailing whitespace, including tabs.
+    assert(Math.abs(Double.parseDouble("6.66D\t ") - 6.66) < epsilon, "a9")
+
+    assert(Math.abs(Double.parseDouble("6.66d") - 6.66) < epsilon, "a10")
+
+    assert(Math.abs(Double.parseDouble("7.77F") - 7.77) < epsilon, "a11")
+    assert(Math.abs(Double.parseDouble("7.77f") - 7.77) < epsilon, "a12")
```
Comment (sjrd) [inline unit-tests/src/test/scala/java/lang/DoubleSuite.scala:208]: `parseDouble` has a very precise definition of its result. We should test it with a strong `==`, not with an `epsilon` difference.

Comment (sjrd) [reply unit-tests/src/test/scala/java/lang/DoubleSuite.scala:208]: Please add tests with a `\u0000` character, such as `"7.77\u0000 "` (should succeed) and `"7.77\u0000a"` (should fail).

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 1 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala:45]: \" not working in interpolated strings is a problem still a problem in many scala versions. It is a problem
in the Scala 2.11 currently used by SN and in the pending Scala 2.12.  I believe that it is unfixed in 
Scala 2.13.2. It may get fixed in 2.13.3 (or .n).

To reduce the donkeying around, I re-wrote the line to achieve the same intent but not rely upon
string interpolation.   I also added an informative section documenting the bug I was working around.

I wanted to steer far clear of octal constants (\042), which work, and Unicode constants (\u0022), which
fail because they are translated too early.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 2 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala:50]: Yes, it is still relevant but looks like it is in the wrong/a_confusing place after exceptionMsg got factored out.
I moved the comment to just before exceptionMsg and documented the bug & need for workaround.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 3 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala:53]: I reworded it.  key change is s/type/value/.   Also tried to remove some mystification  about the return type out of
that if branch.  After a year, I could not follow it, so I figured others might have problems also.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 4 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala:16]: I removed the "Perhaps" line.  I have learned more about Unicode and how Java & Scala define & parse whitespace over the year since this PR was created.   

I followed the valueOf URL you kindly provided and added a reference to 
javase/8/docs/api/java/lang/String.html#trim-- to the comment.

The PR is was explicitly crafted to use the Java String.trim method to ensure consistency. 

Thank you for picking up on the less than stellar comment.  Someone may have to follow this code someday.

File: unit-tests/src/test/scala/java/lang/DoubleSuite.scala (hunk 8 above)
Comment (LeeTibbert) [reply unit-tests/src/test/scala/java/lang/DoubleSuite.scala:208]: In both DoubleSuite.scala & FloatSuite.scala:
* `epsilon` comparisons changed to strong `==` test as recommended.
* Requested tests added to each file.

I also fixed a copy & paste error in FloatSuite.scala where Double.parseDouble() was being called instead of 
the correct Float.parseFloat().

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 5 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala]: I extensively re-worked the implementation based on review comments, sample code, the review 
concern with minimizing allocations and a years worth of experience & understanding.

The idea was to do the math in the C and C byte realm rather than improperly conflating Scala/Java Strings.
I also added a few unit-test tests for corner cases which I found tricky or hard to remember.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 6 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala]: Understood. My mind was in the C world and I did not consider perverse Java string.
 Gone in current iteration.  Thank you for the catch.

File: javalib/src/main/scala/java/lang/IEEE754Helpers.scala (hunk 7 above)
Comment (LeeTibbert) [reply javalib/src/main/scala/java/lang/IEEE754Helpers.scala]: Gone in current iteration.
